### PR TITLE
Cloud Logging support

### LINKED ISSUE #21370: Stackdriver Logging datasource
**What would you like to be added**:

As of now we are able to query and display Stackdriver Metrics into Grafana, I think it would be great to also be able to query and display Stackdriver Logs. I am actually surprised that I was not able to find any issue or pull request about this already.

**Why is this needed**:

This is needed because I want to have my metrics and logs in the same place, making it possible to make dashboards where I could display metrics and logs for a given Stackdriver resource.

### PR DESCRIPTION
**Pull request by mtanda**
**What this PR does / why we need it**:
Add Cloud Logging support.

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/grafana/issues/21370

**Special notes for your reviewer**:
I don' want to stop https://github.com/grafana/grafana/pull/23066 work.
Pick the code from PR.

(Sorry for Erik, I can't resolve code conflict, so I just copy code.)

### CODE REVIEW

File: public/app/plugins/datasource/cloud-monitoring/datasource.ts [hunk 1]
```diff
@@ -132,28 +109,61 @@ export default class CloudMonitoringDatasource extends DataSourceApi<CloudMonito
...
+    const lines = new ArrayVector<string>([]);
+    const ids = new ArrayVector<string>([]);
+
+    for (const entry of data.entries) {
+      times.add(entry.timestamp);
+      lines.add(entry.textPayload);
+      ids.add(entry.insertId);
     }
+
+    const dataFrame = {
```
Comment (mtanda) [inline public/app/plugins/datasource/cloud-monitoring/datasource.ts:149]: set all metadata like https://github.com/mtanda/grafana-google-stackdriver-logging-datasource/blob/master/src/datasource.ts#L126-L161 ?

File: public/app/plugins/datasource/cloud-monitoring/components/LogsQueryEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      <textarea
+        rows={7}
+        value={query.filter}
+        className="gf-form-input gf-form-textarea slate-query-field"
+        onChange={e => onChange({ ...query, filter: e.target.value })}
+        onBlur={onRunQuery}
+        placeholder="Enter a Stackdriver Logging query"
+        required
+      />
+    </>
```
Comment (mtanda) [inline public/app/plugins/datasource/cloud-monitoring/components/LogsQueryEditor.tsx:47]: add pageSize and orderBy field?

**Comment by mtanda** [comment]
I'm waiting for review.

**Comment by mtanda** [comment]
not stale

**Comment by aknuds1** [comment]
Assuming we want this functionality, it sounds like a good idea to also include tests.

**Comment by bergquist** [comment]
Hi mtanda. :wave: 

I'm not sure the maintenance burden is worth it for Grafana Core. Any reason why this couldn't be a plugin instead?

**Comment by mtanda** [comment]
Cloud Logging is core feature of GCP observability.
AWS CloudWatch datasource already support CloudWatch Logs.
I think Cloud Monitoring should support Cloud Logging.

Cloud Logging datasource require to store GCP credential in Grafana backend, and the credential shouldn't be exposed for frontend.
So, Cloud Logging need to be implemented as backend plugin.

Providing Cloud Logging as core datasource is meaningful.
Grafana user can use it without worrying about development stop.
Dashboard is never broken.
Grafana Cloud user can use Cloud Logging datasource.

Hopefully, the code size is small, I think maitenance cost is not much.


Side note:
I believe registering commercial service plugin to grafana-plugin-repository is impossible.
https://github.com/grafana/grafana-plugin-repository/pull/152
https://github.com/grafana/grafana-plugin-repository/pull/660

I can't register it to grafana-plugin-repository, it means most of user doesn't notice the datasource plugin, and I can't share the useful datasource with community (if I want to share).

**Comment by mtanda** [comment]
I'm not sure this PR is acceptable for Grafana core.
When I got answer from Grafana core team, I'll add test.

**Comment by jacobstr** [comment]
We've been inspired for something like this actually. The stackdriver logging UI can be a bit finicky and it would be lovely to build a dashboard with multiple parameterized stackdriver log queries that react to any dashboard filters that have been applied.

E.g. suppose I filter by a kubernetes pod and the panels respond by showing the kubernetes events (stackdriver logging), pod  logs (stackdriver logging) pod resource utilization (in-cluster thanos), for that specific pod. Particularly, constructing these queries tends to be tedious (was it `jsonPayload.labels` or `jsonPayload.metadata.labels`) so the stackdriver experience is creating multiple saved searches opening multiple tabs, substituting in values...

This rings true:

> Cloud Logging is core feature of GCP observability.
> AWS CloudWatch datasource already support CloudWatch Logs.
> I think Cloud Monitoring should support Cloud Logging.

Pardon the "me too" - just wanted to lend some context because I'm coming hot off the heels of thinking this was part of the GrafanaDatasource and was not.

**Comment by mtanda** [comment]
hope to progress this work

**Comment by aknuds1** [comment]
We will look at this this week.

**Comment by mtanda** [comment]
@aknuds1 Thanks lot. I rebased this branch for latest master.

**Comment by aknuds1** [comment]
We're discussing internally if this is something we want to accept.

**Comment by aknuds1** [comment]
@daniellee Can you give some feedback on this one?

**Comment by daniellee** [comment]
@mtanda sorry for the long delay. 

We have a partnership with Google to maintain and improve the Cloud Monitoring data source. When Erik created the original PR, we were waiting to see if they would also support Cloud Logging. Our initial discussions with them about Cloud Logging pointed to lots of potential performance problems. We see this with our other logging data sources as well - it is easy to crash Grafana with queries that return a lot of data and we need to build in logic to prevent this. Before merging support for Cloud Logging, we would like to find a contact person on the Cloud Logging team at Google to help us with that.

So I do not want to merge this until we have support from the Google Cloud Monitoring team. The support burden would be too large. Erik is back from parental leave in December and that would make it possible to do this properly then.

This does not apply to your MQL PR (#26551) which we already have support for from the Google CM team.

**Comment by mtanda** [comment]
@daniellee 

Thanks for reply.

> it is easy to crash Grafana with queries that return a lot of data and we need to build in logic to prevent this.

I understand Grafana core team are concerned about stability.

I agree with the point, Logging datasource could cause many problem.
So, it is better to wait Erik come back.

He already open draft PR, so I close my PR.

**Pull request closed without merging**